lib/helpers.js

```diff
@@ -0,0 +1,49 @@
+const _ = require('lodash');
+
+function resolveAuthLevel(schema, options, doc) {
+  // Look into options the options and try to find authLevels. Always prefer to take
+  // authLevels from the direct authLevel option as opposed to the computed
+  // ones from getAuthLevel in the schema object.
+  let authLevels = [];
+  if (options) {
+    if (options.authLevel) {
+      authLevels = _.castArray(options.authLevel);
+    } else if (typeof schema.getAuthLevel === 'function') {
+      authLevels = _.castArray(schema.getAuthLevel(options.authPayload, doc));
+    }
+  }
+  // Add `defaults` to the list of levels since you should always be able to do what's specified
+  // in defaults.
+  authLevels.push('defaults');
+
+  const perms = schema.permissions || {};
+  return _.chain(authLevels)
+    .filter(level => !!perms[level]) // make sure the level in the permissions dict
+    .uniq() // get rid of fields mentioned in multiple levels
+    .value();
+}
+
+function getAuthorizedFields(schema, options, action, doc) {
+  const authLevels = resolveAuthLevel(schema, options, doc);
+
+  return _.chain(authLevels)
+    .map(level => schema.permissions[level][action])
+    .flatten()
+    .filter(path => schema.path(path) || schema.virtualpath(path)) // make sure the fields are in the schema
```

Comment:
Is this line really needed? I think it is removing nested objects from the list even though they are checked (breaks writing nested objects)

Reply:
Ah, that's a good case to consider. To be clear, you're talking about a schema like this?

```
const schema = Schema({
  foo: {
    bar: String
  }
});
```

And you're saying that you won't be able to set permissions like this?
```
schema.permissions = {
  'foo.bar': { read, write, remove }
};
```

I was torn on adding that extra filter. by having it, the rest of the code was able to not have to worry about bogus paths coming back. Is there a way to validate that paths you are mentioning based on the schema?